client/clientutil/modelinfo.go

```diff
@@ -0,0 +1,461 @@
+
+	// return if verbose is not requested
+	if !options.Verbose {
+		return w.Flush()
+	}
+	return printVerboseAssertionHeaders(w, options, &serialAssertion)
+}
+
+// PrintModelAssertionJSON will format the provided serial or model assertion based on the parameters given in
+// JSON format. The output will be written to the provided io.Writer.
+func PrintModelAssertionJSON(w *tabwriter.Writer, modelFormatter ModelFormatter, options PrintModelAssertionOptions, modelAssertion asserts.Model, serialAssertion *asserts.Serial) error {
+
+	serializeJSON := func(v interface{}) error {
+		marshalled, err := json.MarshalIndent(v, "", "  ")
+		if err != nil {
+			return err
+		}
+		_, err = w.Write(marshalled)
+		w.Flush()
+		return err
+	}
+
+	if options.Assertion {
+		modelJSON := modelAssertJSON{}
+		modelJSON.Headers = modelAssertion.Headers()
+		modelJSON.Body = string(modelAssertion.Body())
+		return serializeJSON(modelJSON)
+	}
+
+	// ordering of the primary keys for model: brand, model, serial
+	// ordering of primary keys for serial is brand-id, model, serial
+
+	// output brand/brand-id
+	modelData := make(map[string]interface{})
+	modelData["brand-id"] = modelAssertion.HeaderString("brand-id")
+	modelData["model"] = modelAssertion.HeaderString("model")
+
+	// for the serial header, if there's no serial yet, it's not an error for
+	// model (and we already handled the serial error above) but need to add a
+	// parenthetical about the device not being registered yet
+	var serial string
+	if serialAssertion == nil {
+		// For JSON verbose is always true, so we need to escape the dash
+		serial = modelFormatter.GetEscapedDash()
+		serial += " (device not registered yet)"
+	} else {
+		serial = serialAssertion.HeaderString("serial")
+	}
+
+	// only output the grade if it is non-empty, either it is not in the model
+	// assertion for all non-uc20 model assertions, or it is non-empty and
+	// required for uc20 model assertions
+	grade := modelAssertion.HeaderString("grade")
+	if grade != "" {
+		modelData["grade"] = grade
+	}
+
+	storageSafety := modelAssertion.HeaderString("storage-safety")
+	if storageSafety != "" {
+		modelData["storage-safety"] = storageSafety
+	}
+
+	// serial is same for all variants
+	modelData["serial"] = serial
+	allHeadersMap := modelAssertion.Headers()
+
+	// always print extra information for JSON
+	for _, headerName := range niceOrdering {
+		invalidTypeErr := fmt.Errorf("invalid type for %q header", headerName)
+
+		headerValue, ok := allHeadersMap[headerName]
+		// make sure the header is in the map
+		if !ok {
+			continue
+		}
+
+		// switch on which header it is to handle some special cases
+		switch headerName {
+		// long base64 key we can rewrap safely
+		case "device-key":
+			headerString, ok := headerValue.(string)
+			if !ok {
+				return invalidTypeErr
+			}
+			// remove the newlines from the string
+			modelData["device-key"] = strings.Replace(headerString, "\n", "", -1)
```

Comment:
Ah, I got you! ;-) There's no split + join here! :-)

Anyway, you can use ReplaceAll instead of passing `-1`.

Reply:
Ah yes! Nice!